File: handlers/downloader.py
# anexpert/handlers/downloader.py
import os
import yt_dlp
import asyncio
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import ContextTypes
from utils.storage_manager import schedule_deletion
from config import Config

async def downloader_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user = update.effective_user
    is_ayah = Config.AYAH_USERNAME and user.username == Config.AYAH_USERNAME
    
    if not context.args:
        if is_ayah:
            msg = (
                "Ayah harus kirimin dulu link videonya yaw -.-\n\n"
                "Caranya:\n"
                "Ketik `/dl [link_video]`\n\n"
                "Jan lupa Yah, berikan Anne jajan sesekali atas kerja kerasnya :v"
            )
        else:
            msg = (
                "Kirimkan link videonya ges\n\n"
                "Caranya:\n"
                "Ketik `/dl [link_video]`"
            )
        await update.message.reply_text(msg, parse_mode="Markdown")
        return
        
    url = context.args[0]
    
    # Simpan URL ke dalam memori sementara pengguna
    context.user_data['last_dl_url'] = url
    
    # Buat tombol pilihan
    keyboard = [
        [
            InlineKeyboardButton("Unduh Video / MP4", callback_data="dl_video"),
            InlineKeyboardButton("Unduh Audio / MP3", callback_data="dl_audio")
        ],
        [InlineKeyboardButton("Batal", callback_data="dl_cancel")]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)

    if is_ayah:
        msg = r"Ayah mau download sebagai apa nih? \(°^°)/"
    else:
        msg = "Pilih format unduhannya yah~"
        
    await update.message.reply_text(msg, reply_markup=reply_markup)

async def downloader_button_callback(update: Update, context: ContextTypes.DEFAULT_TYPE):
    query = update.callback_query
    await query.answer() # Menghentikan loading ikon jam pasir di tombol
    
    user = update.effective_user
    is_ayah = Config.AYAH_USERNAME and user.username == Config.AYAH_USERNAME
    
    data = query.data
    url = context.user_data.get('last_dl_url')
    
    # Jika tombol batal ditekan
    if data == "dl_cancel":
        if is_ayah:
            await query.edit_message_text("Oke Yah, unduhan dibatalkan yaw! :3")
        else:
            await query.edit_message_text("K... Udah batal nih")
        return
        
    if not url:
        await query.edit_message_text("Uhh... Anne udah lupa linknya (¬_¬') coba ketik ulang")
        return

    if is_ayah:
        await query.edit_message_text("Sabar Ayah, Anne lagi ngambil medianya... (｡•̀ᴗ•́)✧")
    else:
        await query.edit_message_text("Tunggu bentar nih, lagi di-download...")
    
    is_audio = (data == "dl_audio")
    
    if is_audio:
        ydl_opts = {
            'format': 'bestaudio/best',
            'outtmpl': 'downloads/%(id)s.%(ext)s',
            'quiet': True,
            'noplaylist': True,
            'postprocessors': [{
                'key': 'FFmpegExtractAudio',
                'preferredcodec': 'mp3',
                'preferredquality': '192',
            }],
        }
    else:
        ydl_opts = {
            'format': 'best[filesize<50M]',
            'outtmpl': 'downloads/%(id)s.%(ext)s',
            'quiet': True,
            'noplaylist': True,
        }
        
    try:
        if not os.path.exists('downloads'):
            os.makedirs('downloads')
            
        def extract_info():
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(url, download=True)
                return ydl.prepare_filename(info)
                
        loop = asyncio.get_event_loop()
        file_path = await loop.run_in_executor(None, extract_info)
        
        # Jika Audio, yt-dlp + ffmpeg akan mengubah ekstensinya menjadi .mp3
        if is_audio:
            base_name = os.path.splitext(file_path)[0]
            mp3_path = f"{base_name}.mp3"
            if os.path.exists(mp3_path):
                file_path = mp3_path
        
        with open(file_path, 'rb') as media_file:
            if is_audio:
                sent_msg = await query.message.reply_audio(audio=media_file)
            else:
                sent_msg = await query.message.reply_video(video=media_file)
            
        # Hapus file setelah 5 menit
        schedule_deletion(context, sent_msg.chat_id, sent_msg.message_id)
        
        await query.message.delete() # Hapus pesan loading
        if os.path.exists(file_path):
            os.remove(file_path) # Bersihkan file lokal

    except Exception as e:
        if is_ayah:
            error_msg = f"M-maaf Ayah, Anne agak bingung (╥﹏╥)... Error teknis pas download: `{str(e)[:100]}`"
        else:
            error_msg = "Maaf nih yek, linknya susah di-download atau file-nya kegedean (maks 50MB)... Coba link lain yah!"
        await query.edit_message_text(error_msg, parse_mode="Markdown")

File: node-gateway/src/handlers/downloader.js
// anexpert/node-gateway/src/handlers/downloader.js
//
// Port dari handlers/downloader.py. Sama seperti versi Python, ini
// memanggil yt-dlp — tapi bukan lewat Python, melainkan lewat BINARY
// standalone yt-dlp (satu file executable, sudah termasuk semua
// dependency-nya, tanpa perlu install Python sama sekali).
//
// Keamanan: kita SELALU pakai execFile dengan argumen berbentuk ARRAY,
// tidak pernah menggabungkan string URL ke command shell, sesuai aturan
// "Sanitasi Batas Antar Proses" di config.md.
const fs = require('fs');
const os = require('os');
const path = require('path');
const crypto = require('crypto');
const { execFile } = require('child_process');
const { isAyah } = require('./start');
const { scheduleDeletion } = require('../utils/storageManager');

function ytdlpBinaryName() {
  return process.platform === 'win32' ? 'yt-dlp.exe' : 'yt-dlp';
}

const YTDLP_BIN = process.env.YTDLP_PATH || path.join(__dirname, '..', '..', 'bin', ytdlpBinaryName());

// Simpan URL terakhir per user secara in-memory (pengganti context.user_data
// versi Python). Untuk bot personal skala kecil ini cukup.
const lastUrlByUser = new Map();

function runYtDlp(args) {
  return new Promise((resolve, reject) => {
    execFile(
      YTDLP_BIN,
      args,
      { maxBuffer: 1024 * 1024 * 50, timeout: 5 * 60 * 1000 },
      (error, stdout, stderr) => {
        if (error) {
          return reject(new Error(stderr ? stderr.slice(-500) : error.message));
        }
        resolve(stdout);
      }
    );
  });
}

async function downloaderHandler(bot, msg, matchArgs) {
  const chatId = msg.chat.id;
  const userId = msg.from.id;
  const ayah = isAyah(msg.from.username);

  const url = (matchArgs || '').trim().split(/\s+/)[0];

  if (!url) {
    const helpMsg = ayah
      ? 'Ayah harus kirimin dulu link videonya yaw -.-\n\n' +
        'Caranya:\nKetik `/dl [link_video]`\n\n' +
        'Jan lupa Yah, berikan Anne jajan sesekali atas kerja kerasnya :v'
      : 'Kirimkan link videonya ges\n\nCaranya:\nKetik `/dl [link_video]`';
    await bot.sendMessage(chatId, helpMsg, { parse_mode: 'Markdown' });
    return;
  }

  lastUrlByUser.set(userId, url);

  const replyMarkup = {
    inline_keyboard: [
      [
        { text: 'Unduh Video / MP4', callback_data: 'dl_video' },
        { text: 'Unduh Audio / MP3', callback_data: 'dl_audio' },
      ],
      [{ text: 'Batal', callback_data: 'dl_cancel' }],
    ],
  };

  const pickMsg = ayah ? 'Ayah mau download sebagai apa nih? \\(°^°)/' : 'Pilih format unduhannya yah~';
  await bot.sendMessage(chatId, pickMsg, { reply_markup: replyMarkup });
}

async function downloaderButtonCallback(bot, query) {
  const chatId = query.message.chat.id;
  const messageId = query.message.message_id;
  const userId = query.from.id;
  const ayah = isAyah(query.from.username);
  const data = query.data;

  await bot.answerCallbackQuery(query.id);

  if (data === 'dl_cancel') {
    const cancelMsg = ayah ? 'Oke Yah, unduhan dibatalkan yaw! :3' : 'K... Udah batal nih';
    await bot.editMessageText(cancelMsg, { chat_id: chatId, message_id: messageId });
    return;
  }

  const url = lastUrlByUser.get(userId);
  if (!url) {
    await bot.editMessageText("Uhh... Anne udah lupa linknya (¬_¬') coba ketik ulang", {
      chat_id: chatId,
      message_id: messageId,
    });
    return;
  }

  const waitMsg = ayah ? 'Sabar Ayah, Anne lagi ngambil medianya... (｡•̀ᴗ•́)✧' : 'Tunggu bentar nih, lagi di-download...';
  await bot.editMessageText(waitMsg, { chat_id: chatId, message_id: messageId });

  const isAudio = data === 'dl_audio';
  const downloadsDir = path.join(os.tmpdir(), 'anexpert-downloads');
  if (!fs.existsSync(downloadsDir)) fs.mkdirSync(downloadsDir, { recursive: true });

  const jobId = crypto.randomUUID();
  const outputTemplate = path.join(downloadsDir, `${jobId}.%(ext)s`);
  
  // Rute absolut mutlak menuju file cookies.txt di dalam folder node-gateway
  const cookiesPath = path.join(__dirname, '..', '..', 'cookies.txt');

  // Argumen selalu berbentuk array — tidak ada penggabungan string shell.
  const args = isAudio
    ? [
        '--cookies', cookiesPath,
        '-f', 'bestaudio/best',
        '--extract-audio',
        '--audio-format', 'mp3',
        '--audio-quality', '192K',
        '--no-playlist',
        '-o', outputTemplate,
        url,
      ]
    : [
        '--cookies', cookiesPath,
        '-f', 'best[filesize<50M]/best',
        '--no-playlist',
        '-o', outputTemplate,
        url,
      ];

  try {
    await runYtDlp(args);

    // Cari file hasil download di folder (ekstensi ditentukan yt-dlp sendiri)
    const files = fs.readdirSync(downloadsDir).filter((f) => f.startsWith(jobId));
    const filePath = isAudio
      ? path.join(downloadsDir, files.find((f) => f.endsWith('.mp3')) || files[0])
      : path.join(downloadsDir, files[0]);

    if (!filePath || !fs.existsSync(filePath)) {
      throw new Error('File hasil download tidak ditemukan.');
    }

    let sentMsg;
    const fileStream = fs.createReadStream(filePath);
    if (isAudio) {
      sentMsg = await bot.sendAudio(chatId, fileStream);
    } else {
      sentMsg = await bot.sendVideo(chatId, fileStream);
    }

    scheduleDeletion(bot, sentMsg.chat.id, sentMsg.message_id);

    await bot.deleteMessage(chatId, messageId).catch(() => {});
    fs.promises.unlink(filePath).catch(() => {});
  } catch (e) {
    const errorMsg = ayah
      ? `M-maaf Ayah, Anne agak bingung (╥﹏╥)... Error teknis pas download: \`${String(e.message).slice(0, 200)}\``
      : 'Maaf nih yek, linknya susah di-download atau file-nya kegedean... Coba link lain yah!';
    await bot.editMessageText(errorMsg, { chat_id: chatId, message_id: messageId, parse_mode: 'Markdown' }).catch(() => {});

    try {
      const leftovers = fs.readdirSync(downloadsDir).filter((f) => f.startsWith(jobId));
      leftovers.forEach((f) => fs.promises.unlink(path.join(downloadsDir, f)).catch(() => {}));
    } catch {}
  }
}

module.exports = { downloaderHandler, downloaderButtonCallback };
